### ISSUE TITLE
bug: Unresolved type reference causes model building failure

### ISSUE THREAD

**Issue by pvojtechovsky** [INRIA/spoon#3087]
Here is failing test https://github.com/pvojtechovsky/spoon/tree/bug3087

**Comment by pvojtechovsky** [comment]
It blocks #2683. I don't know how/have no time to fix it.

**Comment by tdurieux** [comment]
Is  `testCompileUnresolvedFullyQualifiedName` that should fail?
It is passing here :/

**Comment by pvojtechovsky** [comment]
@tdurieux thank you for checking.
You are right. I am sorry. It fails only on #2683, because of some change there. So I will check myself, what is the problem there. 
I am closing this issue

**Comment by tdurieux** [comment]
there are some failing test cases in your branch such as `compileTestWithImportStaticWildcard`.
I have a fix locally, do you need it?

**Comment by tdurieux** [comment]
It seems that the failing test case is related to https://github.com/INRIA/spoon/pull/2683/files#diff-a490aa7d6a5ea006caa9168294d3d9f4R498

`originTypeRef == daikon.PptRelation.PptRelationType`
`typeRef == daikon.PptRelation`

and because of that, the condition at https://github.com/INRIA/spoon/pull/2683/files#diff-a490aa7d6a5ea006caa9168294d3d9f4R511 is false

**Comment by tdurieux** [comment]
My previous message is wrong, sorry.

The reference is equals to:
`originTypeRef == daikon.PptRelation.PptRelationType`

For spoon the package is `daikon` but for jdt, the package is `daikon.PptRelation`.

The test is passing if you add:

```java
if ((packageBinding.tagBits & 128L) != 0L && new String(packageNames[off]).equals(typeRef.getSimpleName())) {
	// Package is missing (tagBits & 128L != 0), potential incorrect packageNames
	off--;
}
```

at https://github.com/INRIA/spoon/pull/2683/files#diff-a490aa7d6a5ea006caa9168294d3d9f4R531

but not completely sure if it is correct

**Comment by pvojtechovsky** [comment]
The problem is debuggable when this PR is rebased on #2683.
It seems to be not related to CtArrayTypeReference.
But the problem is that:
`import static daikon.PptRelation.PptRelationType;`
is unresolved reference and therefore it is not clear whether 
A) `PptRelation` is a package name and `PptRelationType` is a top level class
B) `PptRelation` is a top level class name and `PptRelationType` is a nested class
and Spoon created model where `PptRelation` is a type name, while JDK compiler expects that `PptRelation` is package name. 
Therefore the condition https://github.com/INRIA/spoon/pull/2683/files#diff-a490aa7d6a5ea006caa9168294d3d9f4R530 is not true and it continues to exception.

**Comment by tdurieux** [comment]
Yes, we arrive at the same conclusion (see my previous message).
Do you think that my "fix" is correct?

**Comment by pvojtechovsky** [comment]
@tdurieux it looks like your https://github.com/tdurieux/spoon/tree/CompileUnresolved contains some special fix or feature. I do not understand it's purpose yet, so I will not use it. But please make a PR for that if that makes sense for Spoon.

**Comment by tdurieux** [comment]
I am sorry, I meant this one


```java
if ((packageBinding.tagBits & 128L) != 0L && new String(packageNames[off]).equals(typeRef.getSimpleName())) {
	// Package is missing (tagBits & 128L != 0), potential incorrect packageNames
	off--;
}
```

at https://github.com/INRIA/spoon/pull/2683/files#diff-a490aa7d6a5ea006caa9168294d3d9f4R531



Indeed, https://github.com/tdurieux/spoon/tree/CompileUnresolved  contains something different, it is not throwing an exception when a type is not found in the noclasspath mode.

**Comment by pvojtechovsky** [comment]
> Do you think that my "fix" is correct?
@tdurieux Where is your fix? I hope I did not deleted it by a force push ...

**Comment by pvojtechovsky** [comment]
@tdurieux I committed my fix to #2683, but there is still another problem on #2683 with handling of unresolved references.
ReferenceBuilder line 903
```java
ref.setSimpleName(new String(binding.readableName()));
```
the `binding.readableName()` returns 
`daikon.VarInfo.RefType`
but expected simple name is
`RefType`

I will stop for today. It is your if you have time :-)
